[quant][graphmode] insert_quant_dequant jit pass

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25127 [quant][graphmode] insert_quant_dequant work with qconfig_dict
* #25069 [graphmode] insert_observers use qconfig_dict
* #25000 [cleanup] Remove InsertQuantDeQuantNode
* #24999 [cleanup] Remove insert_observers pass
* #24998 [cleanup] Remove deprecated graph mode quantization tests
* #24428 [quant][graphmode] Integration tests for initial quantization graph mode
* #24427 [quant][graphmode] quant_fusion jit pass
* **#24426 [quant][graphmode] insert_quant_dequant jit pass**

Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

Differential Revision: [D17001141](https://our.internmc.facebook.com/intern/diff/D17001141)

jerryzh168 committed:
[quant][graphmode] insert_quant_dequant jit pass

Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] insert_quant_dequant jit pass"


Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] insert_quant_dequant jit pass"


Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] insert_quant_dequant jit pass"


Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] insert_quant_dequant jit pass"


Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] insert_quant_dequant jit pass"


Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] insert_quant_dequant jit pass"


Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] insert_quant_dequant jit pass"


Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] insert_quant_dequant jit pass"


Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

nishantpdce commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +677,255 @@ TORCH_API script::Module InsertObservers(
+  }
+
+  // quantize bias first
+  for (Value* v : values_to_observe) {
```
Identifying the bias nodes based on iterating the observable nodes and checking whether there is actually not observer and then checking for bias seems indirect approach.

Is there a different way possible. Can this not be a separate pass as it was earlier [quant-dequant for params]?

jerryzh168 replied:
don't we need to do the same in a separate pass?

jerryzh168 replied:
why would we like to split into multiple passes?

jerryzh168 replied:
I think the no-observer should actually be an assert when we find that the bias is an argument for "aten::conv", see "quantizeBias", we can probably move this inside that loop

nishantpdce replied:
i think approach is different now. Earlier parameters did not have observers and so passes were different. Ok to have it done all together. 

Only mechanism to how we get to node is bit indirect.  Since bias should not have observer, we are looking at observable nodes and then checking if no observer module and if its bias seemed indirect to me.

Since bias is unrelated to observers, you can access directly by looking at params to module and mapping them to supported op argument [conv or linear]

jerryzh168 replied:
yeah makes sense, I just make a change to move this check inside quantizeBias

jerryzh168 replied:
we can certainly start from conv/linear and quantize the bias, but I think the current approach is fine too, please take a look again.

nishantpdce commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +677,255 @@ TORCH_API script::Module InsertObservers(
+  }
+}
+
+void quantizeValue(const script::Module& module,
```
can we rename. As we are both quantizing tensor and inserting q-dq node

jerryzh168 replied:
quantizing tensors does require us to calculate qparam for them, I know previously it's separate because we have a global observer module, but in the current setting when we have per tensor observer I think they can be just one step.

nishantpdce replied:
ok sounds good

jerryzh168 committed:
Update on "[quant][graphmode] insert_quant_dequant jit pass"


Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

nishantpdce commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +677,257 @@ TORCH_API script::Module InsertObservers(
+void quantizeBias(const script::Module& module, Value* v) {
+  // Traverse to the place where this is used
+  Node* aten_call_node = nullptr;
+  std::vector<std::string> ops_with_bias = {"aten::conv2d", "aten::_convolution"};
```
how about nn.linear ?

jerryzh168 replied:
will add them later

nishantpdce commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +677,257 @@ TORCH_API script::Module InsertObservers(
+      // Get qparam from weight
+      IValue weight_qparam = getQParam(module, weight);
+      IValue bias_scale =  at::scalar_tensor(
+          c10::Scalar(act_qparam.toTuple()->elements()[0].toTensor().item().toDouble() *
```
Should bias scale not be inverse ? 1 /  (a * w). I recollect from this
https://github.com/pytorch/pytorch/commit/a501e7d5be3845aae3b90fc349eff8e0f2fb05af#diff-78aab5b4217aeaed920efab0c5f5f5c8R1490

jerryzh168 replied:
it should be `act_scale * wt_scale` actually:
`res_q = (x_f * w_f) / (x_s * w_s) + b_f / (x_s * w_s)`, it's probably lenet is not very sensitive to numerical errors previously so that we didn't discover this issue for previous implementation.

nishantpdce replied:
aha got it

jerryzh168 committed:
Update on "[quant][graphmode] insert_quant_dequant jit pass"


Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

nishantpdce commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +677,252 @@ TORCH_API script::Module InsertObservers(
+    // work around until data only clone of module.clone is supported.
+    auto submodule_name = u.user->inputs().at(0)->debugName();
+    if (u.user->kind() == prim::CallMethod && u.user->s(attr::name) == "forward" &&
+//            u.user->inputs().at(0)->type()->kind() == c10::ClassType &&
```
remove this line

nishantpdce reviewed (approved):
One more minor comment related to a commented line. Please remove it before merging

jerryzh168 committed:
Update on "[quant][graphmode] insert_quant_dequant jit pass"


Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

ZolotukhinM commented on test/test_jit.py:
```diff
@@ -1158,6 +1158,60 @@ def get_forward(m):
+                   .check("aten::quantize_linear") \
+                   .check_next("aten::int_repr") \
+                   .check_next("aten::_dequantize_linear") \
+                   .check("aten::conv2d") \
```
This will break after inlining is disabled by default. I suggest porting the test to the final state where it's disabled - I think you would have `prim::CallMethod` nodes here instead of direct aten calls. You could add `@_tmp_donotuse_dont_inline_everything` decorator to the test to disable it for now. Also if you do this there would be no need in running constant-prop.

jerryzh168 replied:
I'll change to final state, since I need this to test qconfig_dict as well

jerryzh168 replied:
it says don't use in the name :), so I won't use it :)

jerryzh168 replied:
oh will probably disable and re-enable later

ZolotukhinM replied:
I think it's fine to use it despite the name. And it'll actually be easier to land the "disable-inlining-by-default" PR if we use it because it's easier to just remove decorators than to fix tests that start to fail.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +677,252 @@ TORCH_API script::Module InsertObservers(
+}
+
+script::Module InsertQuantDeQuant(
+    const script::Module& module,
```
We're mutating the module, it shouldn't be `const`.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +677,252 @@ TORCH_API script::Module InsertObservers(
+  // the module with observer modules as a temporary workaround
+  // TODO: remove observer modules after we have a remove_module API
+
+  return module;
```
I don't know if it makes sense to return module as well when we're mutating it in-place.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +677,252 @@ TORCH_API script::Module InsertObservers(
+    if (u.user->kind() == prim::CallMethod && u.user->s(attr::name) == "forward" &&
+//            u.user->inputs().at(0)->type()->kind() == c10::ClassType &&
+            submodule_name.find("observer_for_") == 0) {
+          auto observer_module = module.find_module(submodule_name);
```
We cannot use `debugName` for any lookups. We need to find `prim::GetAttr` which produced the first argument of the current `prim::CallMethod` and use 'name' attribute of that node.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1,4 +1,5 @@
 #include <torch/csrc/jit/passes/quantization.h>
+#include <torch/csrc/jit/passes/subgraph_rewrite.h>
```
This is not used anywhere.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +677,252 @@ TORCH_API script::Module InsertObservers(
   return input_module;
 }
 
+Node* insertQuantDeQuantCall(Value* v, const IValue& qparams, at::ScalarType t, bool insert_after=true) {
```
Nit: clang-format please :)

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +677,252 @@ TORCH_API script::Module InsertObservers(
+  return dequant;
+}
+
+std::tuple<std::string, c10::optional<script::Module>> findObserverModule(
```
This function have very non-trivial signature. It at least needs to be commented, but might make sense to rethink what it does and returns - I wouldn't expect a `findObject` function to return some string along with the object. It might make sense to split it into to functions if you need to perform two actions (one of which is sometimes ignored as well).

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +677,252 @@ TORCH_API script::Module InsertObservers(
+      return std::make_tuple("", c10::nullopt);
+}
+
+IValue getQParam(const script::Module& module, Value* v) {
```
Almost all methods have the same first argument - module. Please wrap this all in a class and save the module there.

jerryzh168 replied:
this is a big change which might introduce many merge conflict, will do after the whole stack lands.

ZolotukhinM replied:
I'd prefer to do it now, as it simplifies review. And you mostly add code, no changes to the existing one, so the conflicts should be minimal.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +677,252 @@ TORCH_API script::Module InsertObservers(
+
+void quantizeTensor(const script::Module& module,
+                    Value* v,
+                    std::vector<std::string>& observer_modules_to_remove,
```
`const std::vector&`?

ZolotukhinM replied:
nvm

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +677,252 @@ TORCH_API script::Module InsertObservers(
+void quantizeTensor(const script::Module& module,
+                    Value* v,
+                    std::vector<std::string>& observer_modules_to_remove,
+                    std::vector<Node*>& nodes_to_destroy,
```
We could move these two arguments to the class private fields too making signature of this function more comprehensible.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +677,252 @@ TORCH_API script::Module InsertObservers(
+  std::string observer_module_name;
+  c10::optional<script::Module> observer_module;
+  std::tie(observer_module_name, observer_module) = findObserverModule(module, v);
+  if (observer_module) {
```
Nit: if you change it to `if (!observer_module) { return; }` you would save one level of indentation for the entire function :)

ZolotukhinM reviewed (changes requested):
Hi! I haven't finished reviewing yet, but I have put some remarks so that you can start working on them while I'm still reviewing.

jerryzh168 committed:
Update on "[quant][graphmode] insert_quant_dequant jit pass"


Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] insert_quant_dequant jit pass"


Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +676,246 @@ TORCH_API script::Module InsertObservers(
+  Node* insert_point = insert_after ? v->node() : *g->nodes().begin();
+  WithCurrentScope scope_guard(
+      *insert_point->owningGraph(), insert_point->scope());
+  WithInsertPoint ins(insert_point);
```
There is no benefit in using `WithInsertPoint` if we only use explicit `insertBefore`/`insertAfter` calls.

Ideally, we would set the insert point at the callsite of this function and change this function to use `insertNode` in this function. This way we get rid of the `insert_after` argument of this function and make the function branchless.

jerryzh168 replied:
hmm, we still have insertBefore and insertAfter right

jerryzh168 replied:
oh, you mean pass different nodes so that it is always insertAfter/insertBefore?

ZolotukhinM replied:
You don't need to pass any nodes, that's the beauty of `WithInsertPoint` - when you create an object of this type (e.g. `WithInsertPoint x (my_insert_point);`) it sets the global insert-point to `my_insert_point`. This global insert point is used by all `Graph::insert*` methods except `insertBefore` and `insertAfter`. When `WithInsertPoint`objects comes out of the scope, it restores the previous insert point.

So, in this case we could just create an object `WithInsertPoint` before calling this function and then this function will just insert its stuff into the right place.

Yes, you would need to change the insertion logic in function to be like `insertBefore`, but you would call `insertNode` instead (`insertNode` ~ `insertBefore(insert_point)`).

jerryzh168 replied:
oh, you mean it will do insertBefore/insertAfter depending on where the node is?

jerryzh168 replied:
doesn't seem to work, I'll land like this and we can talk about how to refactor later.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +676,246 @@ TORCH_API script::Module InsertObservers(
+  return c10::nullopt;
+}
+
+// TODO: refactor to have a class
```
Could you please add more details to this TODO? It's not clear what exactly you plan to change here in future.

jerryzh168 replied:
done

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +676,246 @@ TORCH_API script::Module InsertObservers(
+
+  // calculate qparams and insert quant/dequant calls
+  auto calculate_qparams = observer_module.value().get_method("calculate_qparams");
+  IValue qparams = calculate_qparams(std::vector<IValue>());
```
Can we use `getQParam` function here? Seems very similar to what it is doing anyway.

jerryzh168 replied:
yeah, actually I can use a flag saying whether we want to remove the observer_module/node to merge the previous code to getQParam as well

ZolotukhinM replied:
I'm not a big fan of adding flags to functions - if we do this, `getQParam` name would no longer reflect what it does, which is not great. It should be renamed to something like `getQParamAndRemoveObserverCall`, which I think would be better split into two separate functions. And this is basically what we currently have (except that the code for `removeObserverCall` isn't factored out to a separate function).

jerryzh168 replied:
ok, it's just we'll findObserverName multiple times in that case, but it should be fine still

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +676,246 @@ TORCH_API script::Module InsertObservers(
+  // remove observer forward call
+  for (const Use& u: v->uses()) {
+    Node* user = u.user;
+    if (user->kind() == prim::CallMethod &&
```
This check can be factored out to a function `isObserverForwardCall` (or something like that), which can be used here and in `findObserverName`.

jerryzh168 replied:
in `findObserverName` we actually don't know the name yet, I think the only check we can extract is `user->kind() == prim::CallMethod && user->s(attr::name) == "forward"` probably not worthwhile to do..

ZolotukhinM commented on torch/csrc/jit/passes/quantization.h:
```diff
@@ -122,5 +122,49 @@ TORCH_API script::Module InsertObservers(
+ */
+TORCH_API void QuantFusion(std::shared_ptr<Graph>& graph);
+
+class QuantizeHelper {
```
Nit: since this class is not for the public use, we should move it to the cpp file.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +676,246 @@ TORCH_API script::Module InsertObservers(
+    // work around until data only clone of module.clone is supported.
+    if (u.user->kind() == prim::CallMethod && u.user->s(attr::name) == "forward") {
+      auto module_instance = u.user->inputs().at(0);
+      if (module_instance->node()->kind() == prim::GetAttr && module_instance->node()->s(attr::name).find("observer_for_") != std::string::npos) {
```
Clang-format the changes please :)

jerryzh168 replied:
plan to do it after to avoid conflicts..

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +676,246 @@ TORCH_API script::Module InsertObservers(
+    for (Node* n : b->nodes()) {
+      // Skip nodes that we don't need to observe, e.g. 'prim::Constant' or
+      // observer nodes
+      if (!outputsNeedToBeObserved(n) || observer_for_input.count(n) != 0) {
```
Doesn't seem that we're putting anything into `observer_for_input`. Do we want to? If not, please remove it.

jerryzh168 replied:
we don't need that, will cleanup

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +676,246 @@ TORCH_API script::Module InsertObservers(
+    const std::string& method_name) {
+  script::Method method = module.get_method(method_name);
+  auto graph = method.graph();
+  TORCH_CHECK(graph != nullptr);
```
I don't think this check could ever fail.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +676,246 @@ TORCH_API script::Module InsertObservers(
+    if (std::find(ops_with_bias.begin(), ops_with_bias.end(),
+                  use.user->kind().toQualString()) != ops_with_bias.end()) {
+      // Make sure there is no observer module for bias
+      c10::optional<script::Module> observer_module;
```
Unused variable?

ZolotukhinM reviewed (approved):
Thanks for addressing my previous remarks! I have only minor cleanup requests and then this is ready to go!

jerryzh168 committed:
Update on "[quant][graphmode] insert_quant_dequant jit pass"


Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +676,246 @@ TORCH_API script::Module InsertObservers(
+void QuantizeHelper::quantizeBias(Value* v) {
+  // Traverse to the place where this is used
+  Node* aten_call_node = nullptr;
+  std::vector<std::string> ops_with_bias = {"aten::conv2d", "aten::_convolution"};
```
when we disable inlining, we also need to go inside forward function of child modules, the current flow assumes that we have a inlined graph.

the change to make it work without inlining will appear in the qconfig_dict PR.

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -676,5 +676,246 @@ TORCH_API script::Module InsertObservers(
+      // Get qparam from weight
+      IValue weight_qparam = getQParam(weight);
+      IValue bias_scale =  at::scalar_tensor(
+          c10::Scalar(act_qparam.toTuple()->elements()[0].toTensor().item().toDouble() *
```
sure, will do

jerryzh168 committed:
Update on "[quant][graphmode] insert_quant_dequant jit pass"


Summary:
Added following pass:
- _jit_pass_insert_quant_dequant: removes observer modules and calls, insert
quantize_linear-int_repr-_dequantize_linear calls for activation, weight and bias,
the scale of bias is calculated from the scale of input activation and weight

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:
